Decouple Keystone from `ApiCredentials`

mokagio opened the pull request:
Builds on top of https://github.com/wordpress-mobile/WordPress-iOS/pull/24435

mokagio committed:
Inject `ApiCredentials` `.client` and `.secret`

That's instead of doing a direct read under the hood.

The aim of this is to get to a place where Keystone is agnostic of
`ApiCredentials`.

mokagio committed:
Inject `ApiCredentials` Google client details

That's instead of doing a direct read under the hood.

The aim of this is to get to a place where Keystone is agnostic of
`ApiCredentials`.

mokagio committed:
Inject `ApiCredentials` Tenor API key

That's instead of doing a direct read under the hood.

The aim of this is to get to a place where Keystone is agnostic of
`ApiCredentials`.

mokagio committed:
Inject `ApiCredentials` Sentry DSN

That's instead of doing a direct read under the hood.

The aim of this is to get to a place where Keystone is agnostic of
`ApiCredentials`.

mokagio committed:
Inject `ApiCredentials` `encryptedLogKey`

That's instead of doing a direct read under the hood.

The aim of this is to get to a place where Keystone is agnostic of
`ApiCredentials`.

mokagio committed:
Inject `ApiCredentials` `docsBotId`

That's instead of doing a direct read under the hood.

The aim of this is to get to a place where Keystone is agnostic of
`ApiCredentials`.

mokagio committed:
Define secrets type in BuildSettingsKit

mokagio committed:
Add secrets property to `BuildSettings`

mokagio committed:
Configure secrets in `BuildSettings` using `ApiCredentials`

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/308">25.9</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio committed:
Replace accessing secrets from `ApiCredentials` with `BuildSettings`

mokagio committed:
Remove `Secrets.swift` from Keystone, add to Reader

mokagio committed:
Add placeholder in script to generate credentials for Reader

mokagio commented on Modules/Sources/BuildSettingsKit/BuildSecrets.swift:
```diff
@@ -0,0 +1,100 @@
+public struct BuildSecrets: Sendable {
```
Right now, this is no different from `ApiCredentials` (`Secrets.swift`). Eventually, it'll be the interface we use to inject the obfuscated encrypted secrets.

I considered having a dedicated framework for this, but it seemed a waste given BuildSettingsKit already had all the plumbing to access the values.

I also considered putting it in Keystone, but it seems to me having this in BuildSettingsKit will help us extract to smaller more modules in the future.

mokagio commented on Modules/Sources/BuildSettingsKit/BuildSecrets.swift:
```diff
@@ -0,0 +1,100 @@
+        case .preview:
+            return .dummy
+        case .test:
+            // TODO: Should we crash if a secret is accessed from the tests to prevent under-the-hood access and favor injection?
```
Keen to hear what you think of this.

mokagio replied:
Otherwise we can keep it like this. It might be interesting to see if new tests fail because they run with an empty secret.

kean replied:
If the tests don't use these, I would add a `fatalError`. The only reason I removed it from `BuildSettings.current` in general was because there were too many tests using it and it was not feasible to rewrite them.

mokagio commented on Scripts/BuildPhases/GenerateCredentials.sh:
```diff
@@ -52,6 +52,12 @@ ensure_is_in_input_files_list $WORDPRESS_SECRETS_FILE
 JETPACK_SECRETS_FILE="${SECRETS_ROOT}/Jetpack-Secrets.swift"
 ensure_is_in_input_files_list $JETPACK_SECRETS_FILE
 
+# TODO: We'll need a dedicated one for Reader...
+# READER_SECRETS_FILE="${SECRETS_ROOT}/Reader-Secrets.swift"
+# ...using WordPress-Secrets.swift for now.
+READER_SECRETS_FILE="${SECRETS_ROOT}/WordPress-Secrets.swift"
+ensure_is_in_input_files_list $READER_SECRETS_FILE
```
Just waiting for the appropriate entries to be created. See also p1744279985523659/1744233208.119539-slack-C08LE4U7FAM

mokagio commented on WordPress/Classes/System/ApiCredentials+BuildSecrets.swift:
```diff
@@ -0,0 +1,25 @@
+            debuggingKey: debuggingKey
+        )
+    }
+}
```
This is temporary. If/when we'll have obfuscation, we'll do without `ApiCredentials` (`Secrets.swift`) and use dedicated instances of `BuildSecrets` for each of the apps.

mokagio commented on WordPress/WordPress.xcodeproj/project.pbxproj:
```diff
@@ -111,7 +111,6 @@
 		0C5A8A812D9B22F100C25301 /* WordPressAuthenticator.framework in Embed Frameworks */ = {isa = PBXBuildFile; fileRef = 4AD953B42C21451700D0EEFA /* WordPressAuthenticator.framework */; settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
 		0C5A8A842D9B22F100C25301 /* yoga.xcframework in Frameworks */ = {isa = PBXBuildFile; fileRef = 3F60D3932D2C4BA4008ACD86 /* yoga.xcframework */; };
 		0C5A8A852D9B22F100C25301 /* yoga.xcframework in Embed Frameworks */ = {isa = PBXBuildFile; fileRef = 3F60D3932D2C4BA4008ACD86 /* yoga.xcframework */; settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
-		0C5C46F62D9837CE00F2CD55 /* Secrets.swift in Sources */ = {isa = PBXBuildFile; fileRef = 24351253264DCA08009BB2B6 /* Secrets.swift */; };
```
The diff is not entirely clear, but what's going on here is:

- The derived source `Secrets.swift` went from Keystone to Reader
- So did the build phase to generate it.

mokagio commented on WordPress/Classes/Login/WordPressDotComAuthenticator.swift:
```diff
@@ -77,8 +77,8 @@ struct WordPressDotComAuthenticator {
         redirectURIScheme: String = BuildSettings.current.appURLScheme,
-        clientId: String = ApiCredentials.client,
-        clientSecret: String = ApiCredentials.secret
+        clientId: String = BuildSettings.current.secrets.oauth.client,
+        clientSecret: String = BuildSettings.current.secrets.oauth.secret
```
This might is not the time, but if we keep the nested types (`BuildSecrets.OAuth` in this instance) then we can also pass the type itself here. It would keep the signature simpler.

mokagio committed:
Do not upload encrypted logs during unit tests

After migrating to reading secrets from `BuildSecrets`, unit tests
started crashing in the background thread attempting the encrypted logs
upload.

This is possibly related to the changes in how the secrets are accessed,
but anyways we should not upload logs from the unit tests as that's just
noise.

mokagio commented on WordPress/Classes/Utility/Logging/EventLoggingDelegate.swift:
```diff
@@ -1,10 +1,15 @@
     var shouldUploadLogFiles: Bool {
+        guard BuildSettingsEnvironment.current != .test else {
+            return false
+        }
```
After migrating to reading secrets from `BuildSecrets`, unit tests
    started crashing in the background thread attempting the encrypted logs
    upload.

    This is possibly related to the changes in how the secrets are accessed,
    but anyways we should not upload logs from the unit tests as that's just
    noise.

kean replied:
Nice! It also used  to crash for OSS setup.

crazytonyli commented on Modules/Sources/BuildSettingsKit/BuildSecrets.swift:
```diff
@@ -0,0 +1,100 @@
+public struct BuildSecrets: Sendable {
+
+    public struct OAuth: Sendable {
```
This is the WordPress.com app secret, right? What do you think about naming this to be dotcom specific, instead of a generic term "OAuth"?

mokagio replied:
Good point. 

I'll merge this PR as is, but will followup. Maybe `DotComOAuth`? Or, given the one below is `Google`, we could use `WordPressDotCom` or just `WPCom`?

mokagio committed:
Add demo pipeline that builds Reader for ASC distribution (#24440)

mokagio committed:
Merge remote-tracking branch 'origin/trunk' into mokagio/build-settings-secrets-wrapper

kean commented on config/Reader.debug.xcconfig:
```diff
@@ -1,6 +1,8 @@
-#include "Common.alpha.xcconfig"
+#include "Common.debug.xcconfig"
```
mb 👀

dcalhoun commented:
@Gio2018 this appears to have introduced https://github.com/wordpress-mobile/WordPress-iOS/issues/24562. Will you please help me determine an appropriate solution? 🙇🏻‍♂️

dcalhoun commented:
Just learned this was already being addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/24565. Sorry for the unnecessary ping.
